plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcTypeTranslator.java

```diff
@@ -197,6 +212,65 @@ private OrcTypeTranslator() {}
             };
         }
 
+        if (fromOrcType.getOrcTypeKind() == STRUCT && toTrinoType instanceof RowType rowType) {
+            ImmutableList.Builder<Optional<TypeCoercer<? extends Type, ? extends Type>>> coercers = ImmutableList.builder();
+            ImmutableList.Builder<RowType.Field> fromField = ImmutableList.builder();
+            ImmutableList.Builder<RowType.Field> toField = ImmutableList.builder();
+
+            List<String> fromStructFieldName = fromOrcType.getFieldNames();
+            List<String> toStructFieldNames = rowType.getFields().stream()
+                    .map(RowType.Field::getName)
+                    .map(Optional::get)
+                    .collect(toImmutableList());
+
+            for (int i = 0; i < toStructFieldNames.size(); i++) {
+                if (i >= fromStructFieldName.size()) {
+                    toField.add(new RowType.Field(
+                            Optional.of(toStructFieldNames.get(i)),
+                            rowType.getFields().get(i).getType()));
+                    coercers.add(Optional.empty());
+                }
+                else {
+                    Optional<TypeCoercer<? extends Type, ? extends Type>> coercer = createCoercer(
+                            nestedColumns.get(i).getColumnType(),
+                            nestedColumns.get(i).getNestedColumns(),
+                            rowType.getFields().get(i).getType());
+                    coercers.add(coercer);
+
+                    Type rowFieldType = rowType.getFields().get(i).getType();
+                    fromField.add(new RowType.Field(
+                            Optional.of(fromStructFieldName.get(i)),
+                            coercer.map(TypeCoercer::getFromType).orElse(rowFieldType)));
+                    toField.add(new RowType.Field(
+                            Optional.of(toStructFieldNames.get(i)),
+                            coercer.map(TypeCoercer::getToType).orElse(rowFieldType)));
+                }
+            }
+
+            return Optional.of(new StructCoercer(RowType.from(fromField.build()), RowType.from(toField.build()), coercers.build()));
```

Comment:
What happens when none of the fields need a coercion ? Do we still get overhead of going through the wrapper ?
If coercion is not supported for some of the fields, do they turn into NULLs or do we get an exception ?

Reply:
> Do we still get overhead of going through the wrapper
This I think can we can avoid it - In case of coercer for partitioned tables - we could return `Optional#empty` we can try the same thing here. 

> If coercion is not supported for some of the fields, do they turn into NULLs or do we get an exception ?

If the coercion is not supported - then we would return `Optional#empty` so the coercion will not be applied - if the underlying type doesnt match with table data type - it would throw an error.